packages/vuelidate/src/index.js

```diff
@@ -61,67 +43,100 @@ export function useVuelidate (validations, state, registerAs) {
   // provide to all of it's children the remove results  function
   provide(VuelidateRemoveChildResults, removeChildResultsFromParent)
 
-  const validationResults = computed(() => setValidations({
-    validations: unwrap(validations),
-    state,
-    childResults,
-    resultsCache
-  }))
+  return { childResults, sendValidationResultsToParent, removeValidationResultsFromParent }
+}
+
+/**
+ * Composition API compatible Vuelidate
+ * Use inside the `setup` lifecycle hook
+ * @param {Object|null} validations - Validations Object
+ * @param {Object} state - State object
+ * @param {String} globalConfig - Config Object
+ * @return {UnwrapRef<*>}
+ */
+export function useVuelidate (validations, state, globalConfig = {}) {
+  let { $registerAs } = globalConfig
+
+  const instance = getCurrentInstance()
+
+  // if there is no registration name, add one.
+  if (!$registerAs) {
+    // NOTE:
+    // ._uid // Vue 2.x Composition-API plugin
+    // .uid // Vue 3.0
+    const uid = instance.uid || instance._uid
+    $registerAs = `_vuelidate_${uid}`
+  }
+  const validationResults = ref({})
+  const resultsCache = new ResultsStorage()
+
+  const { childResults, sendValidationResultsToParent, removeValidationResultsFromParent } = nestedValidations()
+
+  // Options API
+  if (!validations && instance.type.validations) {
+    const rules = instance.type.validations
+
+    state = ref({})
+    onBeforeMount(() => {
+      // Delay binding state to validations defined with the Options API until mounting, when the data
+      // has been attached to the component instance. From that point on it will be reactive.
+      state.value = instance.proxy
+
+      // helper proxy for instance property access. It makes every reference
+      // reactive for the validation function
+      function ComputedProxyFactory (target) {
+        return new Proxy(target, {
+          get (target, prop, receiver) {
+            return (typeof target[prop] === 'object')
+              ? ComputedProxyFactory(target[prop])
+              : computed(() => target[prop])
+          }
+        })
+      }
+
+      watch(() => isFunction(rules) ? rules.call(instance.proxy, new ComputedProxyFactory(instance.proxy)) : rules,
+        (validations) => {
+          validationResults.value = setValidations({
+            validations,
+            state,
+            childResults,
+            resultsCache,
+            globalConfig
+          })
+        }, { immediate: true })
+    })
+
+    globalConfig = instance.type.validationsConfig || {}
+  } else {
+    const validationsWatchTarget = isRef(validations)
+      ? validations
+      : () => validations
+
+    watch(validationsWatchTarget, (newValidationRules) => {
```

Comment:
That’s the thing! Even though we rerun the whole `setValidations` if the results are cached (where possible), we use those instead of creating new computed properties.